pyo3-ffi/src/listobject.rs

```diff
@@ -64,3 +66,15 @@ extern "C" {
     #[cfg_attr(PyPy, link_name = "PyPyList_SET_ITEM")]
     pub fn PyList_SET_ITEM(arg1: *mut PyObject, arg2: Py_ssize_t, arg3: *mut PyObject);
 }
+
+#[inline]
+#[cfg(not(Py_NOGIL))]
+pub unsafe fn PyList_FetchItem(list: *mut PyObject, index: Py_ssize_t) -> *mut PyObject {
+    _Py_XNewRef(PyList_GetItem(list, index))
+}
+
+#[inline]
+#[cfg(Py_NOGIL)]
+pub unsafe fn PyList_FetchItem(list: *mut PyObject, index: Py_ssize_t) -> *mut PyObject {
+    _PyList_FetchItem(list, index)
+}
```

Comment:
TBH, adding these provisional APIs and changing PyO3 to use them is a 👎 from me, the callsites which you modified should _already_ be converting borrowed references to owned, and if they didn't, that's a separate bug.

Reply:
The idea behind this proposed change is that, when running without the GIL, the retrieval from list and the acquisition of an owned reference needs to be performed atomically in relation to any concurrent modifications. The PEP 703 proposes `PyList_FetchItem` and `PyDict_GetItem` which have this behavior -- they return owned references and are safe in the face of concurrent modification.